### PR TITLE
Move test_multi_tensor_optimizers to use OptimizerInfos

### PR DESCRIPTION
**Pull request by janeyx99**
This PR aims for parity+ compared to the old testing for the simplest foreach test case.

Test coverage increase: we now test foreach optimizers with CPU as well as on GPU.

Before:
```
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (19136605)]$ python test/test_optim.py -v -k test_multi_tensor_optimizers
/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
test_multi_tensor_optimizers (optim.test_optim.TestOptim) ... ok

----------------------------------------------------------------------
Ran 1 test in 7.253s

OK
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (19136605)]$
```

Now, we get granular test cases at the cost of overhead!
```
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (19136605)]$ python test/test_optim.py -v -k test_foreach
/home/janeyx/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
test_foreach_ASGD_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_Adadelta_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_Adagrad_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_AdamW_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_Adam_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_Adamax_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_NAdam_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_RAdam_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_RMSprop_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_Rprop_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_SGD_cpu_float64 (__main__.TestOptimRenewedCPU) ... ok
test_foreach_ASGD_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_Adadelta_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_Adagrad_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_AdamW_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_Adam_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_Adamax_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_NAdam_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_RAdam_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_RMSprop_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_Rprop_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok
test_foreach_SGD_cuda_float64 (__main__.TestOptimRenewedCUDA) ... ok

----------------------------------------------------------------------
Ran 22 tests in 30.954s

OK
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (19136605)]$ 
```

Why the increase in time?
Two reasons:
1. overhead. Any _CUDA_ *Info test (OpInfo, ModuleInfo, OptimizerInfo) will wrap itself with the `CudaNonDefaultStream` policy, and `CudaNonDefaultStream.__enter__` when called for the first time will go through all visible CUDA devices and synchronize each of them, thus forcing the CUDAContext to be init'd. Doing this for all 8 devices takes ~10-15s. Also, test parametrization costs a little overhead too, but not to the level init'ing CUDA context does.
2. We test more! Now, we have 72 configs (in the foreach optimizer world) whereas we only had 59 before.

Next steps for the future:
- consider adding more Tensor LR configs (like a Tensor LR without capturable in the single tensor case)
- this is likely the next PR or 2: migrate all uses of _test_derived_optimizers in test_optim to TestOptimRenewed

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115299
* #115025
* #115023
* #114802
* __->__ #114797

### CODE REVIEW

**Comment by janeyx99** [comment]
> Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
> 
> * **->** [Move test_multi_tensor_optimizers to use OptimizerInfos #114797](https://github.com/pytorch/pytorch/pull/114797)
> * [Introduce OptimizerInfos + add a test_errors #114178](https://github.com/pytorch/pytorch/pull/114178)
> * [[BE][SparseAdam] cleaner way to verify no sparse params #114425](https://github.com/pytorch/pytorch/pull/114425)

File: test/test_optim.py [hunk 1]
```diff
@@ -29,6 +31,78 @@ def test_errors(self, device, dtype, optim_info):
...
+                optimizer = optim_cls(params, **kwargs)
+
+                for i in range(kIterations):
+                    optimizer.zero_grad()
+                    output = model(input)
+                    loss = output.sum()
+                    loss.backward()
+
+                    # Test that step behaves as expected (a no-op) when grads are set to None
+                    if i == 0:
```
Comment (albanD) [inline test/test_optim.py:81]: nit: maybe i == 3 would be better? If this is the very first iteration, there are no momentum buffers anyways. So we wouldn't see the difference?

File: torch/testing/_internal/common_optimizers.py [hunk 2]
```diff
@@ -183,6 +184,32 @@ def test_wrapper(*args, **kwargs):
                 raise ex
 
 
+# ----------------------------------------------------------------------------------------------------------------
+# NOTE: The following optim_inputs_func_* sampling functions only return constructor combinations of NON-IMPLEMENTATION
+# -CHANGING flags, i.e., flags that are not foreach, fused, or differentiable. The idea is that
```
Comment (albanD) [inline torch/testing/_internal/common_optimizers.py:189]: nit: bad line breaking

File: torch/testing/_internal/common_optimizers.py [hunk 3]
```diff
@@ -183,6 +184,32 @@ def test_wrapper(*args, **kwargs):
                 raise ex
 
 
+# ----------------------------------------------------------------------------------------------------------------
+# NOTE: The following optim_inputs_func_* sampling functions only return constructor combinations of NON-IMPLEMENTATION
```
Comment (albanD) [inline torch/testing/_internal/common_optimizers.py:188]: Define what these flags are!, in particular I see maximize below which I would expect to be one such flag :D 
Also let's have a test that check that these flags are never provided in the sample inputs to ensure this is followed

Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py]: i have done so!

File: torch/testing/_internal/common_optimizers.py [hunk 4]
```diff
@@ -183,6 +184,52 @@ def test_wrapper(*args, **kwargs):
...
+# NOTE: [optimizer kwarg categories]
+# We categorize optimizer kwargs as 3 types:
+#  1. optimizer-specific flags are like amsgrad or rho or beta, flags that are specific to
+#     algorithms and thus only show up for certain optimizers. There are many of these, so I
+#     do not bother gathering them all and listing them here. The converse to these would be
+#     global flags that every optimizer ideally _should_ support. We break global flags into
+#     2 further categories and list them all below.
+#  2. global-friendly = ["lr", "weight_decay", "maximize", "capturable"]
+#     global-friendly flags are global flags who play nicely with all other global flags,
+#     i.e., are mutually exclusive in function. This means that any pair of the following
```
Comment (albanD) [inline torch/testing/_internal/common_optimizers.py:197]: Feels like this i.e. says the opposite of the rest of this paragraph?

**Review by albanD — APPROVED** [review]
SGTM

File: torch/testing/_internal/common_optimizers.py (hunk 4 above)
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:197]: mutually exclusive in function doesn’t mean exclusive in the social sense tho

**Pull request closed without merging**